### PR TITLE
[LAA] Don't require Stride == 1/-1 for inbounds pointer AddRecs nowrap.

### PR DESCRIPTION
**Pull request by fhahn**
I might be missing something, but I *think* the checks for Stride == 1/-1 may be more restrictive than necessary.

If we have a pointer AddRec, the maximum increment is 2^(pointer-index-wdith - 1) - 1. This means that if incrementing the AddRec wraps, the distance between the previously accessed location and the wrapped location is > 2^(pointer-index-wdith - 1), i.e. if the GEP for the AddRec is inbounds, this would be poison due to the object being larger than half the pointer index type space. The poison would be immediate UB when the memory access gets executed..

Similar reasoning can be applied for decrements I think.

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
See https://reviews.llvm.org/D149936 for a previous implementation of the same change. We agreed there that this is correct.

The comment could use an update to stop mentioning "unit stride".

It would probably be cleaner to check hasNoUnsignedSignedWrap() instead of isInBounds(), because then this would directly map to the property that LAA is looking for.

**Comment by fhahn** [comment]
> See https://reviews.llvm.org/D149936 for a previous implementation of the same change. We agreed there that this is correct.
> 
> The comment could use an update to stop mentioning "unit stride".
> 

Ah, the comment update was not included in the pushed PR, should be update now!


> It would probably be cleaner to check hasNoUnsignedSignedWrap() instead of isInBounds(), because then this would directly map to the property that LAA is looking for.

Sounds good, there's already https://github.com/llvm/llvm-project/pull/112223

File: llvm/test/Transforms/LoopVectorize/AArch64/sve-gather-scatter.ll [hunk 1]
```diff
@@ -286,55 +286,58 @@ define void @gather_nxv4i32_ind64_stride2(ptr noalias nocapture %a, ptr noalias
 ; CHECK-NEXT:  entry:
 ; CHECK-NEXT:    [[TMP0:%.*]] = call i64 @llvm.vscale.i64()
 ; CHECK-NEXT:    [[TMP1:%.*]] = shl nuw nsw i64 [[TMP0]], 3
-; CHECK-NEXT:    [[MIN_ITERS_CHECK:%.*]] = icmp ult i64 [[N:%.*]], [[TMP1]]
+; CHECK-NEXT:    [[MIN_ITERS_CHECK:%.*]] = icmp ugt i64 [[N:%.*]], [[TMP1]]
```
Comment (artagnon) [inline llvm/test/Transforms/LoopVectorize/AArch64/sve-gather-scatter.ll:289]: Seems like a strange change: can you verify that this is correct?

**Review by artagnon — COMMENTED** [review]
I wanted to investigate this myself, but it looks like you've beaten me to it. Apart from requiring a rebase, tightening of the commit message, and one strange change, I think the reasoning makes sense to me.

File: llvm/test/Transforms/LoopVectorize/AArch64/sve-gather-scatter.ll (hunk 1 above)
Comment (fhahn) [reply llvm/test/Transforms/LoopVectorize/AArch64/sve-gather-scatter.ll]: I double checked what's going on and the successors of the block have been swapped, but the names for the patterns for the blocks haven't been updated automatically (only where they have been used). Updated the pattern names. There's still a change from `ult` to `ule`, which is due to us now vectorizing with interleave groups, which requires the scalar epilogue to execute at least once.

**Comment by fhahn** [comment]
ping :)

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**